File: 105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.py
'''
因为先序中的节点为根节点,所以取出先序的第一个节点
用先序的第一个节点可以将中序分成左右子树，然后取出先序的第二个节点将左右子树再次划分
当将中序全部划分为单个点时就结束
开始递归的时候用了列表作为参数，结果内存溢出，改用下标作参数，先序和中序序列作为全局变量。

'''

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        self.preorder = preorder
        self.inorder = inorder
        return self._buildTree(0, len(preorder), 0, len(inorder))

    def _buildTree(self, pre_start, pre_end, in_start, in_end):
        if pre_start >= pre_end:
            return None
        root = TreeNode(self.preorder[pre_start])
        offset = self.inorder[in_start:in_end + 1].index(root.val)
        root.left = self._buildTree(pre_start + 1, pre_start + offset + 1, in_start, in_start + offset)
        root.right = self._buildTree(pre_start + offset + 1, pre_end, in_start + offset + 1, in_end)
        return root

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.java
/**
方法来源：https://blog.csdn.net/buppt/article/details/78998568
首先我们要知道前序遍历是先访问根节点，再访问左子树，再访问右子树。中序遍历先访问左子树，再访问根节点，最后访问右子树。

例如一棵树如下：

       A
   B       C
 D   E   F   G
1
2
3
前序遍历结果是A BDE CFG 
中序遍历结果是DBE A FCG

我用空格分开了根节点、左子树和右子树。然后我们可以发现规律：

前序遍历的第一个节点是根节点，然后可以根据前序遍历和中序遍历找到左子树和右子树，然后递归分别求左子树和右子树的子树根节点即可。


*/
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer, Integer> inMap = new HashMap<Integer, Integer>();//这里用hashmap比循环找要快。
        for(int i=0;i<inorder.length;i++){
            inMap.put(inorder[i],i);
        }
        return build(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1, inMap);
    }
    public TreeNode build(int[] preorder, int preStart, int preEnd, int[] inorder, int inStart, int inEnd, Map<Integer, Integer> inMap){
        if(preStart>preEnd||inStart>inEnd){
            return null;
        }
        //前序遍历的第一个节点即该子树的根节点。
        TreeNode root=new TreeNode(preorder[preStart]);
        int inRoot=inMap.get(root.val);
        int numsLeft=inRoot - inStart;
        //然后我们就可以知道两个子树的起点和终点
        root.left = build(preorder, preStart + 1, preStart + numsLeft, inorder, inStart, inRoot - 1, inMap);
        root.right = build(preorder, preStart + numsLeft + 1, preEnd, inorder, inRoot + 1, inEnd, inMap);
        return root;
    }
}
